TensorIterator: Check for memory overlap in all `binary_op`s

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43423 Check for internal memory overlap in some indexing-type functions
* #43422 TensorIteratorConfig: Check memory overlap by default
* #43421 TensorIterator: Check for memory overlap in all `nullary_op`s
* #43420 TensorIterator: Check for memory overlap in all `compare_op`s
* **#43419 TensorIterator: Check for memory overlap in all `binary_op`s**
* #43418 TensorIterator: Check memory overlap in all `unary_op`s

Differential Revision: [D23298655](https://our.internmc.facebook.com/intern/diff/D23298655)

peterbell10 committed:
TensorIterator: Check for memory overlap in all `binary_op`s

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `binary_op`s"




[ghstack-poisoned]

zou3519 commented on test/test_nn.py:
```diff
@@ -11910,6 +11910,13 @@ def test_leaky_relu_inplace_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             F.leaky_relu_(x)
 
+    def test_threshold_inplace_overlap(self, device):
+        x = torch.randn((1, 6), device=device).expand((6, 6))
+        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
+            F.threshold(x, 0.5, 0.5, inplace=True)
+        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
+            F.threshold_(x, 0.5, 0.5)
```
Actually, in-place threshold is one of the cases where we can get BC-breaking behavior, right? The following would fail after this PR:
```
import torch
import torch.nn.functional as F
x = F.threshold(torch.tensor([1., -1]).expand(5, 2), threshold=0, value=0.5, inplace=True)
```
So just like `fill_` I think we should have an exception for in-place threshold

zou3519 reviewed (commented):
We should not change `threshold_` to check memory overlap (that looks like it is introduced in this PR due to the default of `binary_op` changing) because it is idempotent to avoid breaking BC (just like for `fill_`). But otherwise this looks good to me

zou3519 reviewed (approved):
pre-emptive approve to get some (internal) things moving along

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `binary_op`s"


Differential Revision: [D23298655](https://our.internmc.facebook.com/intern/diff/D23298655)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `binary_op`s"


Differential Revision: [D23298655](https://our.internmc.facebook.com/intern/diff/D23298655)

[ghstack-poisoned]

zou3519 commented:
It looks like XLA is just going through a different mechanism for all of these.

> Aug 26 00:39:36 ======================================================================
Aug 26 00:39:36 FAIL [0.005s]: test_index_add_mem_overlap_xla (__main__.TestTorchDeviceTypeXLA)
Aug 26 00:39:36 ----------------------------------------------------------------------
Aug 26 00:39:36 Traceback (most recent call last):
Aug 26 00:39:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 257, in instantiated_test
Aug 26 00:39:36     result = test(self, *args)
Aug 26 00:39:36   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 13350, in test_index_add_mem_overlap
Aug 26 00:39:36     x.index_add_(0, ind, value)
Aug 26 00:39:36 AssertionError: RuntimeError not raised
Aug 26 00:39:36 
Aug 26 00:39:36 ======================================================================
Aug 26 00:39:36 FAIL [0.003s]: test_shift_mem_overlap_xla (__main__.TestTorchDeviceTypeXLA)
Aug 26 00:39:36 ----------------------------------------------------------------------
Aug 26 00:39:36 Traceback (most recent call last):
Aug 26 00:39:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 257, in instantiated_test
Aug 26 00:39:36     result = test(self, *args)
Aug 26 00:39:36   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 13355, in test_shift_mem_overlap
Aug 26 00:39:36     x[:-1] <<= x[1:]
Aug 26 00:39:36 AssertionError: RuntimeError not raised
Aug 26 00:39:36 
Aug 26 00:39:36 ----------------------------------------------------------------------

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `binary_op`s"


Differential Revision: [D23298655](https://our.internmc.facebook.com/intern/diff/D23298655)

[ghstack-poisoned]

zou3519 commented:
When I tried landing this, some internal code told me there was a merge conflict. I don't see github reporting a merge conflict, though :/. @peterbell10 could you rebase this please and I'll try to land it again?

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `binary_op`s"


Differential Revision: [D23298655](https://our.internmc.facebook.com/intern/diff/D23298655)

[ghstack-poisoned]

peterbell10 commented:
I think `ghstack` hides conflicts because the PR is submitted against `gh/peterbell10/5/base` and not `master`.
